### ISSUE TITLE
MDB starts 2 minutes late

### ISSUE THREAD

**Issue by yoshi742** [eclipse-ee4j/glassfish#23315]
## Environment Details
* GlassFish Version (and build number): 5.1
* JDK version: 8
* OS: Linux
* Database: SQL Server

----------

## Problem Description
* beahavior
  * current behavior
    * MDB starts 2 minutes late.
  * expected behavior
    * MDB starts within 10 seconds.
* State: easily reproducible

No errors before or after MDB startup.

## Steps to reproduce
1. set DB
    1. create DB
        * DB name=`exampledb`
    1. create DB user.
        * User/Password=`exampleuser/examplepass`
    1. set `Server Roles` for `exampleuser`.
        * select all server roles.
    1. set `User Mapping` for `exampleuser`.
        * select all user mapping for `exampledb`.
1. start message broker.
    ```
    # imqbrokerd -name TestBroker -port 17696
    [#|2021-01-05T18:23:55.124+0900|WARNING|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|[S2004]: Log output channel com.sun.messaging.jmq.util.log.SysLogHandler is disabled: no imqutil in java.library.path|#]


    [#|2021-01-05T18:23:55.139+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|
    ================================================================================
    Message Queue 5.1.3
    Oracle
    Version:  5.1.3  (Build 1-a)
    Compile:  January 27 2019 1625

    Copyright (c) 2013, 2018 Oracle and/or its affiliates.  All rights reserved.
    ================================================================================
    Java Runtime: 1.8.0_222-ea Oracle Corporation /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.222.b03-1.el7.x86_64/jre
    |#]


    [#|2021-01-05T18:23:55.155+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|   IMQ_HOME=/opt/glassfish5/mq
    |#]


    [#|2021-01-05T18:23:55.156+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|IMQ_VARHOME=/opt/glassfish5/glassfish/domains/domain1/imq
    |#]


    [#|2021-01-05T18:23:55.157+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|Linux 3.10.0-1062.el7.x86_64 amd64 VM037644.localdomain (1 cpu) root
    |#]


    [#|2021-01-05T18:23:55.158+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|Java Heap Size: max=190080k, current=190080k
    |#]


    [#|2021-01-05T18:23:55.159+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|Arguments: -name TestBroker -port 17696
    |#]


    [#|2021-01-05T18:23:55.159+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|Broker Properties: imq.instancename=TestBroker, imq.portmapper.port=17696
    |#]


    [#|2021-01-05T18:23:55.207+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|[B1060]: Loading persistent data...
    |#]


    [#|2021-01-05T18:23:55.208+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|Using built-in file-based persistent store: /opt/glassfish5/glassfish/domains/domain1/imq/instances/TestBroker/
    |#]


    [#|2021-01-05T18:23:55.439+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|[B1270]: Processing messages from transaction log file...
    |#]


    [#|2021-01-05T18:23:55.756+0900|FORCE|5.1.3|imq.log.Logger|_ThreadID=1;_ThreadName=main;|[B1039]: Broker "TestBroker@VM037644.localdomain:17696" ready.
    |#]
    ```
1. change `imq.imqcmd.password`
    ```
    # vi /opt/glassfish5/mq/etc/passfile.sample

    - imq.imqcmd.password=myadminpassword
    + imq.imqcmd.password=admin
    ```
1. set JDBC driver
    ```
    # cp mssql-jdbc-<version>.jre8.jar /opt/glassfish5/glassfish/domains/domain1/lib/
    # touch /opt/glassfish5/glassfish/domains/domain1/config/domain.xml
    # asadmin stop-domain
    Waiting for the domain to stop .
    Command stop-domain executed successfully.
    # asadmin start-domain
    Waiting for domain1 to start .........
    Successfully started the domain : domain1
    domain  Location: /opt/glassfish5/glassfish/domains/domain1
    Log File: /opt/glassfish5/glassfish/domains/domain1/logs/server.log
    Admin Port: 4848
    Command start-domain executed successfully.
    ```
1. set GlassFish
    ```
    # asadmin create-jdbc-connection-pool --datasourceclassname com.microsoft.sqlserver.jdbc.SQLServerDataSource --restype javax.sql.DataSource --property User=exampleuser:Password=examplepass:DatabaseName=exampledb:PortNumber=1433:ServerName=<DB-server-IP-address> sqlserverNonXAConnectionPool1
    JDBC connection pool sqlserverNonXAConnectionPool1 created successfully.
    Command create-jdbc-connection-pool executed successfully.
    # asadmin create-jdbc-resource --connectionpoolid sqlserverNonXAConnectionPool1 --enabled=true --target server --property User=test jdbc/NonXADataSourceDB1
    JDBC resource jdbc/NonXADataSourceDB1 created successfully.
    Command create-jdbc-resource executed successfully.
    # 
    # imqcmd create dst -b localhost:17696 -n MDBPhysicalQueue -t q -u admin -passfile /opt/glassfish5/mq/etc/passfile.sample
    Warning [A2016]: Unencoded name=value entry found in passfile /opt/glassfish5/mq/etc/passfile.sample. Please run 'imqusermgr encode' before use the passfile. Support of passfile that has unencoded name=value entries will be removed in a future release.

    Creating a destination with the following attributes:

    Destination Name    MDBPhysicalQueue
    Destination Type    Queue

    On the broker specified by:

    -------------------------
    Host         Primary Port
    -------------------------
    localhost    17696

    Successfully created the destination.

    # asadmin create-jms-resource --restype javax.jms.Queue --property Name=MDBPhysicalQueue --target server jms/MDBQueue
    Administered object jms/MDBQueue created.
    Command create-jms-resource executed successfully.
    # asadmin create-jms-resource --restype javax.jms.QueueConnectionFactory --property tester=bob --target server jms/MDBQueueConnectionFactory
    Connector resource jms/MDBQueueConnectionFactory created.
    Command create-jms-resource executed successfully.
    # 
    # asadmin create-resource-adapter-config --property ConnectionFactoryProperties=imqBrokerHostPort\\=17696:SupportsXA=false:ProviderIntegrationMode=javabean:ConnectionFactoryClassName=com.sun.messaging.ConnectionFactory:\\CommonSetterMethodName=setProperty:UserName=guest:Password=guest:\\QueueConnectionFactoryClassName=com.sun.messaging.QueueConnectionFactory:\\TopicConnectionFactoryClassName=com.sun.messaging.TopicConnectionFactory:\\QueueClassName=com.sun.messaging.Queue:TopicClassName=com.sun.messaging.Topic:\\LogLevel=INFO:UserName=guest:Password=guest:XAQueueConnectionFactoryClassName=com.sun.messaging.XAQueueConnectionFactory:\\XATopicConnectionFactoryClassName=com.sun.messaging.XATopicConnectionFactory:\\XAConnectionFactoryClassName=com.sun.messaging.XAConnectionFactory genericra_local
    Command create-resource-adapter-config executed successfully.
    # 
    # asadmin deploy --dropandcreatetables=false --precompilejsp=false --target server --verify=false --force=true --name genericra_local ~/work/genericra.rar
    Application deployed with name genericra_local.
    Command deploy executed successfully.
    # 
    # asadmin create-connector-connection-pool --steadypoolsize 8 --maxpoolsize 32 --maxwait 60000 --poolresize 2 --idletimeout 300 --raname genericra_local --connectiondefinition javax.jms.QueueConnectionFactory --transactionsupport LocalTransaction --isconnectvalidatereq=false --property fnst=fnst qcpool_local
    Connector connection pool qcpool_local created.
    Command create-connector-connection-pool executed successfully.
    # asadmin create-connector-resource --target server --poolname qcpool_local jms/QCFactory_local
    Connector resource jms/QCFactory_local created.
    Command create-connector-resource executed successfully.
    # 
    # asadmin create-admin-object --raname genericra_local --restype javax.jms.Queue --target server --property DestinationProperties=imqDestinationName\\=clientQueue_local jms/clientQueue_local
    Administered object jms/clientQueue_local created.
    Command create-admin-object executed successfully.
    # 
    # asadmin deploy --dropandcreatetables=false --precompilejsp=false --target server --verify=false --force=true --name TranUseCaseEAR ~/work/TranUseCaseEAR.ear
    Application deployed with name TranUseCaseEAR.
    Command deploy executed successfully.
    ```

1. stop DAS
    ```
    # asadmin stop-domain
    ```
1. start DAS
    ```
    # asadmin start-domain
    ```
1. access `http://localhost:8080/TranUseCaseJSP/?test=Test1`
    * The following response is received 10 seconds after the access. This is because the MDB does not start within 10 seconds.
        ```
        Test1 : FAILED
        ```
    * The application also prints the following message.
        ```
        ERROR - ### TEST Test1 : FAILED! 
        ```

    * The following message is output by the application 2 minutes after access.
        ```
        MDB: received message ...
        ```

[reproduce.zip](https://github.com/eclipse-ee4j/glassfish/files/5770756/reproduce.zip)

## Impact of Issue
MDB starts 2 minutes late. Probably 2 minutes is EJB Container timeout.

[//]: # "Please choose one each Type, Component, and Priority label"

## addition
The following stack trace is output at DAS startup. This stack trace is similar to  https://github.com/eclipse-ee4j/glassfish/issues/21629.
```
  JTS5043: Unexpected error occurred while creating transaction manager instance
MultiException stack 1 of 1
java.lang.IllegalStateException: Circular dependency involving org.glassfish.resourcebase.resources.listener.ResourceManager was found.  Full descriptor is SystemDescriptor(
        implementation=org.glassfish.resourcebase.resources.listener.ResourceManager
        name=ResourceManager
        contracts={org.glassfish.resourcebase.resources.listener.ResourceManager,org.jvnet.hk2.config.ConfigListener}
        scope=org.glassfish.hk2.runlevel.RunLevel
        qualifiers={}
        descriptorType=CLASS
        descriptorVisibility=NORMAL
        metadata=runLevelValue={2},runLevelMode={0},Bundle-SymbolicName={org.glassfish.main.resourcebase.resources.nucleus-resources},Bundle-Version={5.1.0}
        rank=0
        loader=OsgiPopulatorPostProcessor.HK2Loader(OSGiModuleImpl:: Bundle = [org.glassfish.main.resourcebase.resources.nucleus-resources [218]], State = [READY],744342969)
        proxiable=null
        proxyForSameScope=null
        analysisName=null
        id=35
        locatorId=0
        identityHashCode=1487381851
        reified=true)
        at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:204)
        at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:65)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:92)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:66)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getAllServices(ServiceLocatorImpl.java:837)
        at com.sun.jts.jta.TransactionServiceProperties.initRecovery(TransactionServiceProperties.java:294)
        at com.sun.jts.jta.TransactionManagerImpl.<init>(TransactionManagerImpl.java:185)
        at com.sun.jts.jta.TransactionManagerImpl.getTransactionManagerImpl(TransactionManagerImpl.java:153)
        at com.sun.enterprise.transaction.jts.JavaEETransactionManagerJTSDelegate.postConstruct(JavaEETransactionManagerJTSDelegate.java:121)
        at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:303)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:351)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:59)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:47)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:74)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:131)
        at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:176)
        at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:98)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetAllServiceHandles(ServiceLocatorImpl.java:1468)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getAllServices(ServiceLocatorImpl.java:795)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getAllServices(ServiceLocatorImpl.java:783)
        at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.initDelegates(JavaEETransactionManagerSimplified.java:1480)
        at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.postConstruct(JavaEETransactionManagerSimplified.java:141)
        at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:303)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:351)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:59)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:47)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:74)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:131)
        at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:176)
        at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:98)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:92)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:675)
        at org.jvnet.hk2.internal.ThreeThirtyResolver.resolve(ThreeThirtyResolver.java:54)
        at org.jvnet.hk2.internal.ClazzCreator.resolve(ClazzCreator.java:188)
        at org.jvnet.hk2.internal.ClazzCreator.resolveAllDependencies(ClazzCreator.java:211)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:334)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:59)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:47)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:74)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:131)
        at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:176)
        at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:98)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:662)
        at org.jvnet.hk2.internal.ThreeThirtyResolver.resolve(ThreeThirtyResolver.java:54)
        at org.jvnet.hk2.internal.ClazzCreator.resolve(ClazzCreator.java:188)
        at org.jvnet.hk2.internal.ClazzCreator.resolveAllDependencies(ClazzCreator.java:211)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:334)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:59)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:47)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:74)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:131)
        at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:176)
        at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:98)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:754)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getUnqualifiedService(ServiceLocatorImpl.java:766)
        at org.jvnet.hk2.internal.IterableProviderImpl.get(IterableProviderImpl.java:87)
        at com.sun.enterprise.resource.deployer.ResourceAdapterConfigDeployer.getConnectorRuntime(ResourceAdapterConfigDeployer.java:75)
        at com.sun.enterprise.resource.deployer.ResourceAdapterConfigDeployer.deployResource(ResourceAdapterConfigDeployer.java:62)
        at org.glassfish.resourcebase.resources.listener.ResourceManager.deployResources(ResourceManager.java:159)
        at org.glassfish.resourcebase.resources.listener.ResourceManager.postConstruct(ResourceManager.java:105)
        at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:303)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:351)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:281)
        at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:65)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:92)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:66)
        at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.oneJob(CurrentTaskFuture.java:1213)
        at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.run(CurrentTaskFuture.java:1144)
        at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$UpOneLevel.run(CurrentTaskFuture.java:762)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
]]
```

**Comment by hs536** [comment]
This problem is probably due to a deadlock caused by a circular dependency between the TransactionManager and the ResourceManager and that may have been caused by the following commit[*1]:
[*1] https://github.com/javaee/glassfish/commit/4e8ae2867cb2fcd971920ff71b83e9056ecc07eb#diff-3066aeed99c72ac022adf1fe4c3caa266a60194b58fc82d6ae43e4a5b23910e8

The fix seems to follow the Payara fix[*2], but the offending line appears to have been removed from Payara[*3].
[*2] https://github.com/eclipse-ee4j/glassfish/issues/21175#issuecomment-422029217
[*3] https://github.com/payara/Payara/pull/223/commits/e60bab015e4fa4140309631bc1018b5731b792ed

This fix should work for GlassFish as well, what do you think?

**Comment by gerdogdu** [comment]
@hs536 Can we update the GF code as in [*3] payara/Payara@e60bab0?

**Comment by smillidge** [comment]
Fixed by #23325
